File: src/main/java/com/linebot/service/user/BotUserService.java
package com.linebot.service.user;

import com.linebot.entity.BotUser;
import com.linebot.repository.BotUserRepository;
import com.linebot.util.Utils;
import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import javax.validation.constraints.NotNull;
import java.util.List;

@Slf4j
@AllArgsConstructor
@Service
public class BotUserService {
    private BotUserRepository botUserRepository;

    @Transactional(readOnly = true)
    public List<BotUser> findActiveUser() {
        return botUserRepository.findByDeleted("0");
    }

    @Transactional
    public BotUser insert(@NotNull String userId) {
        BotUser u = new BotUser();
        u.setUserId(userId);
        u.setDeleted("0");
        u.setCreatedDate(Utils.now());
        return botUserRepository.save(u);
    }

    @Transactional
    public BotUser delete(@NotNull String userId) {
        BotUser user = botUserRepository.findByUserId(userId);
        if (user == null) {
            return null;
        }
        user.setDeleted("1");
        return botUserRepository.save(user);
    }
}

File: src/main/kotlin/com/linebot/service/user/BotUserService.kt
package com.linebot.service.user

import com.linebot.entity.BotUser
import com.linebot.repository.BotUserRepository
import com.linebot.util.RandomUtils
import com.linebot.util.Utils
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class BotUserService(
    private val botUserRepository: BotUserRepository
) {

    @Transactional(readOnly = true)
    fun findActiveUser(): List<BotUser> {
        return botUserRepository.findByDeleted("0")
    }

    @Transactional
    fun insert(userId: String): BotUser {
        return botUserRepository.save(BotUser(
                userId = userId,
                appUserId = RandomUtils.createUserId(),
                deleted = "0",
                createdDate = Utils.now()
        ))
    }

    @Transactional
    fun delete(userId: String): BotUser? {
        return botUserRepository.findByUserId(userId)?.let {
            it.deleted = "1"
            botUserRepository.save(it)
        }
    }

    @Transactional
    fun getOrCreate(userId: String): BotUser {
        return botUserRepository.findByUserId(userId) ?: insert(userId)
    }
}
